simplesat/sat/policy.py

```diff
@@ -1,39 +1,179 @@
+            (integer > 0) and values are one of (True, False, None).
+        clauses : List of Clause
+            The collection of Clause objects to satisfy.
         """
 
 
 class DefaultPolicy(IPolicy):
+
+    def __init__(self, *args):
+        pass
+
+    def add_requirements(self, assignments):
+        pass
+
     def get_next_package_id(self, assignments, _):
         # Given a dictionary of partial assignments, get an undecided variable
         # to be decided next.
-        undecided = [
+        undecided = (
             package_id for package_id, status in six.iteritems(assignments)
             if status is None
-        ]
-        return undecided[0]
+        )
+        return next(undecided)
+
+
+class PolicyLogger(IPolicy):
 
+    def __init__(self, PolicyFactory, pool, installed_repository):
+        self._policy = PolicyFactory(pool, installed_repository)
+        self._log_installed = list(installed_repository.iter_packages())
+        self._log_suggestions = []
+        self._log_required = []
+        self._log_pool = pool
+        self._log_assignment_changes = []
+
+    def get_next_package_id(self, assignments, clauses):
+        self._log_assignment_changes.append(assignments.get_changelog())
+        pkg_id = self._policy.get_next_package_id(assignments, clauses)
+        self._log_suggestions.append(pkg_id)
+        return pkg_id
+
+    def add_requirements(self, package_ids):
+        self._log_required.extend(package_ids)
+        self._policy.add_requirements(package_ids)
+
+    def _log_histogram(self, pkg_ids):
+        c = Counter(pkg_ids)
+        lines = (
+            "{:>25} {}".format(self._log_pretty_pkg_id(k), v)
+            for k, v in sorted(c.items(), key=lambda p: p[1])
+        )
+        pretty = '\n'.join(lines)
+        return c, pretty
+
+    def _log_pretty_pkg_id(self, pkg_id):
+        p = self._log_pool._id_to_package[pkg_id]
+        return '{} {}'.format(p.name, p.version)
+
+
+class _InstalledFirstPolicy(IPolicy):
+
+    CURRENT = int(-2e7)
+    REQUIRED = int(-1e7)
+
+    def __init__(self, pool, installed_repository):
+        self._pool = pool
+        self._installed_pkg_ids = set(
+            pool.package_id(package) for package in
+            installed_repository.iter_packages()
+        )
+        self._seen_pkg_ids = set()
+        self._required_pkg_ids = set()
+        self._pkg_id_priority = {}
+        self._unassigned_pkg_ids = PriorityQueue()
+
+    def _update_pkg_id_priority(self, assignments=None):
+        pkg_id_priority = {}
+        if assignments is not None:
+            self._seen_pkg_ids.update(assignments.keys())
+
+        ordered_pkg_ids = sorted(
+            self._seen_pkg_ids,
+            key=lambda p: self._pool._id_to_package[p].version,
+            reverse=True,
+        )
+
+        for priority, pkg_id in enumerate(ordered_pkg_ids):
+
+            # Determine the new priority of this pkg
+            if pkg_id in self._required_pkg_ids:
```

Comment:
If I understand correctly the logic, we have one priority queue divided in 3 non-overlapping bands, whose boundaries are defined w/ `REQUIRED` and `CURRENT`. What's the rationale to use this instead of 3 priority queues ?